File: src/adventofcode/day8.py
from collections import deque


def init_screen(screen, rect_str):
    rect_str = rect_str.split(' ')[-1]
    x, y = rect_str.strip().split('x')
    x, y = int(x), int(y)
    for i in range(y):
        for j in range(x):
            screen[i][j] = '#'


def rotate(screen, rot_str):
    s = rot_str.split()[1:]
    idx = int(s[1].split('=')[-1])
    by = int(s[-1])
    if s[0] == 'row':
        screen[idx].rotate(by)
    else:
        dq = deque([i[idx] for i in screen])
        dq.rotate(by)
        for i, j in zip(screen, dq):
            i[idx] = j


def count_pixels(screen):
    pixels = 0
    for row in screen:
        pixels += row.count('#')
    return pixels


if __name__ == '__main__':
    data = ""

    screen = [deque(50 * '.') for _ in range(6)]
    for row in data.split('\n'):
        row = row.strip()
        if not row:
            continue
        elif row.startswith('rect'):
            init_screen(screen, rect_str=row)
        else:
            rotate(screen, rot_str=row)

    print(count_pixels(screen))
    for row in screen:
        print(''.join(row))

File: src/adventofcode/day8.clj
(ns adventofcode.day8
  (:require [clojure.string :as str]))

(defn parse [line]
  (let [nums (map read-string (re-seq #"\d+" line))]
    (cond
      (.startsWith line "rect") [:rect nums]
      (.startsWith line "rotate row") [:rotate-row nums]
      (.startsWith line "rotate col") [:rotate-col nums])))

(def input (->> (slurp "resources/day8_input")
                str/split-lines
                (map parse)))

(def blank-screen (vec (repeat 6 (vec (repeat 50 :off)))))

(defn rect [screen [n-col n-row]]
  (reduce (fn [m coord]
            (assoc-in m coord :on))
          screen
          (for [x (range n-col)
                y (range n-row)]
            [y x])))

(defn rotate-row [screen [y n]]
  (let [row (cycle (nth screen y))
        width (count (first screen))
        shift (- width (rem n width))]
    (->> (drop shift row)
         (take width)
         vec
         (assoc screen y))))

(defn transpose [screen]
  (apply (partial mapv vector) screen))

(defn rotate-col [screen [col n]]
  (-> (transpose screen)
      (rotate-row [col n])
      transpose))

(defn draw [m [instruction coords]]
  (case instruction
    :rect (rect m coords)
    :rotate-col (rotate-col m coords)
    :rotate-row (rotate-row m coords)))

(defn swipe [screen instructions]
  (reduce draw screen instructions))

;; Part 2

(defn print-screen [screen]
  (let [pixels (flatten screen)]
    (println (count (filter #{:on} pixels)))
    (->> pixels
         (map #(case %
                 :on "#"
                 :off " "))
         (partition 50)
         (map str/join)
         (map println))))

(print-screen (swipe blank-screen input))
